### PR TITLE
lantiq: enlarge AVM Fritz!Box 3370 flash

### PR DESCRIPTION
**Pull request by mkuron**
Increase the available flash memory size in AVM Fritz!Box 3370 by
incorporating the unused extra partitions located after the ubi partition.

Available flash space for rootfs+overlay increases from 48MB to 124MB.

Reverting to the OEM firmware is still possible (via the recovery utility
provided by AVM) as the OEM firmware appears to reformat the _config_ and
_nand-filesystem_ partitions upon first boot if necessary. The
_reserved-kernel_ and _reserved-filesystem_ partitions are overwritten by the
OEM firmware when installing an update, so their contents do not matter.

Boot loader and device-specific information (MAC addresses, calibration
data, etc.) are not located in NAND flash and remain unharmed by this
changed.

Tested with OEM firmware 06.54 on device with HWRevision 5 and Micron
flash chip.

@mkresin, can you have a look?

### CODE REVIEW

**Comment by mkresin** [comment]
I kept the original partition layout, as it would allow us to implement a dual image layout.

But it would need r/w tffs access and most likely tffs access during boot (in kernel) + custom mtd code to select the current active partition. I don't see that anyone is implementing r/w tffs, tffs in kernel isn't really something I like, neither I like more custom kernel patches.

To my knowledge the dual boot support in the EVA bootloader is rather rudimentary. No detection of a failed boot and no way to forcefully switch the active partition at power on (via button or similar).

Long story short: Fine with me, if you add a note to the commit message that user having any prior OpenWrt version installed need to do a ftp install via EVA to switch to new partition layout. Doing a sysupgrade from an image using the old partition layout to an image with the new partition layout should cause issue, since the extra space isn't formatted as ubi.

**Comment by mkuron** [comment]
> Doing a sysupgrade from an image using the old partition layout to an image with the new partition layout should cause issue, since the extra space isn't formatted as ubi.

I switched to the new layout via sysupgrade and it worked fine. I can check again though just to be safe. Doesn’t the overlay ubifs get reformatted anyway during sysupgrade? If the rootfs size of the upgrade is different than the installed one, the offset of the ubifs partition changes, so it could hardly be reused.

**Comment by mkuron** [comment]
> Doing a sysupgrade from an image using the old partition layout to an image with the new partition layout should cause issue, since the extra space isn't formatted as ubi.

Oh really, is that necessary? I always thought that during a sysupgrade, the squashfs image was flashed to the beginning of the rootfs partition and a new empty overlay partition was created right after it to fill the rest of the space of rootfs. Since the size of the squashfs can vary, overlays can't be preserved between upgrades. Is that any different now that we are using UBI between the MTD and the actual file systems?

**Comment by mkresin** [comment]
> Oh really? I always thought that during a sysupgrade, the squashfs image was flashed to the beginning of the rootfs mtd partition and a new empty overlay partition was created right after it to fill the rest of the space of rootfs.

Your description is correct for board with NOR flash with data on raw flash. But the fritzbox uses NAND and ubi.

> Is that any different now that we are using UBI between the MTD and the actual file systems?

It works completely different. The following is technically completely wrong but should make it easier to get how things are working.

Ubi can be seen as an own partition system. During installation via EVA, a special image is written to the partition named "ubi", which creates the ubi structure (ubiformat) within the partition.

An ubi structure/partition has sub partitions (volumes). During sysupgrade the volumes are removed and recreated with the content of the sysupgrade image. But the ubi structure/partition is neither touched nor formated during sysupgrade.

After sysupgrading from an image with the small ubi partition to an image with the large ubi partition, the ubi subsystem should be upset on boot, since the ubi ubi structure/partition created during the initial EVA installation ends in the middle of nowhere.

It might work, but I would expect issues. It might be that ubi is nowday able to format the (new) extra space, but I havn't read something like that so far.

**Comment by mkresin** [comment]
> I don't think this is a huge issue as Fritz!Box 3370 support was merged after the 18.06 release was branched. So there has never been a release version supporting this device and not many people will be sysupgrading it.

I know, that is why I'm fine to merge the change.
